content/blog/2024/removing-mixins.md

```diff
@@ -0,0 +1,375 @@
+
+Let's go back to our `visually-hidden` example.
+We should really clarify that
+we don't want it applied when an element has focus.
+With the custom property,
+we can override a single property
+wherever necessary:
+
+```css
+[visually-hidden] {
+  --hidden: var(--hidden--on);
+}
+
+[hidden-when=small] {
+  @container (inline-size < 20ch) {
+    --hidden: var(--hidden--on);
+  }
+}
+
+:focus,
+:active {
+  --hidden: var(--hidden--off);
+}
+```
+
+But with `@apply` we either have to
+plan ahead for all conditions
+_before we apply the mixin_:
+
+```css
+[visually-hidden]:not(:focus):not(:active) {
+  @apply --visually-hidden;
+}
+```
+
+Or we need to carefully revert
+every property of the mixin.
+But… revert _to what value_?
+Maybe we can use `revert-layer`
+and some clever layering?
+We could even build the off-switch into our mixin:
+
+```css
+/* when '--off' is undefined we get the output */
+@mixin --visually-hidden(--off) {
+  clip: var(--off, rect(0 0 0 0));
+  clip-path: var(--off, inset(50%));
+  height: var(--off, 1px);
+  overflow: var(--off, hidden);
+  position: var(--off, absolute);
+  white-space: var(--off, nowrap);
+  width: var(--off, 1px);
+}
+
+@layer base {
+  [visually-hidden] {
+    @apply --visually-hidden;
+  }
+}
+
+/* put our overrides in a higher layer to revert from */
+@layer overrides {
+  :focus,
+  :active {
+    /* set --off to revert-layer */
+    @apply --visually-hidden(revert-layer);
+  }
+}
+```
+
+That provides an off-switch,
+but it requires some careful planning ahead,
+and the layering requirement seems fragile.
+
+## Should mixin calls cascade?
+
+This isn't a new or theoretical issue.
+After years of using Sass mixins,
+it's something I've encountered many times.
+In most situations
+it's possible to work around the issue,
+but sometimes it becomes quite complicated
+to get all the logic right in one place.
+
+The cascade is useful for these situations.
+When we're defining new CSS features
+we often ask _should these behaviors cascade together_,
+are they intertwined?
+If so, they belong in the same property.
+
+It seems clear to me that it would be
+useful (at least sometimes)
+for the mixin-application syntax to cascade.
+But before we make any big changes
+we also need to ask:
+
+- Are there place we _don't want_ that behavior?
+  Places we want to call a mixin twice with different arguments,
+  and have both apply?
+  I haven't thought of good examples,
+  but they might exist?
+- Can we design a cascading mixin-application syntax?
+  We intentionally avoided cascading mixin _definitions_ --
+  one mixin name can't refer to different things in different places.
+  Would cascading `@apply` rules have similar issues?
+
+I don't think we want something like
+the JavaScript `removeEventListener()` function,
+which requires a careful matching of arguments.
+I'd like to avoid any `@un-apply`-style rules.
+That has always seemed fragile to me,
+and I'd rather use the cascade mechanic if we can.
```

Comment:
should "mechanic" be "mechanism"?

Reply:
Maybe? Or it might work without that word entirely?

Reply:
```suggestion
and I'd rather use the cascade if we can.
```